src/vs/editor/contrib/dropIntoEditor/browser/dropProgressWidget.ts

```diff
@@ -0,0 +1,147 @@
+import { IPosition } from 'vs/editor/common/core/position';
+import { Range } from 'vs/editor/common/core/range';
+import { IEditorDecorationsCollection } from 'vs/editor/common/editorCommon';
+import { TrackedRangeStickiness } from 'vs/editor/common/model';
+import { ModelDecorationOptions } from 'vs/editor/common/model/textModel';
+import { localize } from 'vs/nls';
+import { IInstantiationService } from 'vs/platform/instantiation/common/instantiation';
+
+const dropIntoEditorProgress = ModelDecorationOptions.register({
+	description: 'drop-into-editor-progress',
+	stickiness: TrackedRangeStickiness.NeverGrowsWhenTypingAtEdges,
+	showIfCollapsed: true,
+	after: {
+		content: noBreakWhitespace,
+		inlineClassName: 'drop-into-editor-progress-decoration',
+		inlineClassNameAffectsLetterSpacing: true,
+	}
+});
+
+interface DropProgressDelegate {
+	cancel(): void;
+}
+
+class InlineDropProgressWidget extends Disposable implements IContentWidget {
+	private static readonly ID = 'editor.widget.dropProgressWidget';
+
+	allowEditorOverflow = false;
+	suppressMouseDown = true;
+
+	private domNode!: HTMLElement;
+
+	constructor(
+		private readonly editor: ICodeEditor,
+		private readonly range: Range,
+		private readonly delegate: DropProgressDelegate,
+	) {
+		super();
+
+		this.create();
+
+		this.editor.addContentWidget(this);
+		this.editor.layoutContentWidget(this);
+	}
+
+	private create(): void {
+		this.domNode = dom.$('.inline-drop-progress-widget');
+		this.domNode.role = 'button';
+		this.domNode.title = localize('dropIntoEditorProgress', "Running drop handlers. Click to cancel");
+
+		const iconElement = dom.$('span.icon');
+		this.domNode.append(iconElement);
+
+		iconElement.classList.add(...ThemeIcon.asClassNameArray(Codicon.loading), 'codicon-modifier-spin');
+
+		const updateSize = () => {
+			const lineHeight = this.editor.getOption(EditorOption.lineHeight);
+			this.domNode.style.height = `${lineHeight}px`;
+			this.domNode.style.width = `${Math.ceil(0.8 * lineHeight)}px`;
+		};
+		updateSize();
+
+		this._register(this.editor.onDidChangeConfiguration(c => {
+			if (c.hasChanged(EditorOption.fontSize) || c.hasChanged(EditorOption.lineHeight)) {
+				updateSize();
+			}
+		}));
+
+		this._register(dom.addDisposableListener(this.domNode, dom.EventType.CLICK, e => {
+			this.delegate.cancel();
+		}));
+	}
+
+	getId(): string {
+		return InlineDropProgressWidget.ID;
+	}
+
+	getDomNode(): HTMLElement {
+		return this.domNode;
+	}
+
+	getPosition(): IContentWidgetPosition | null {
+		return {
+			position: { lineNumber: this.range.startLineNumber, column: this.range.startColumn },
+			preference: [ContentWidgetPositionPreference.EXACT]
+		};
+	}
+
+	override dispose(): void {
+		super.dispose();
+		this.editor.removeContentWidget(this);
+	}
+}
+
+export class DropProgressManager extends Disposable {
+
+	/** Delay before showing the progress widget */
+	private readonly _showDelay = 500; // ms
```

Comment:
I think the equivalent default delay is 150ms in notification progress, any reason to deviate? https://github.com/mjbvz/vscode/blob/b45c925dd0296392abf6ab3911c207fa32924b1d/src/vs/workbench/services/progress/browser/progressService.ts#L82

Reply:
This widget shifts around the line text so we want to be cautious about showing it too quickly. That's why I went with half a second instead